plugins/out_s3/s3.c

```diff
@@ -252,6 +253,143 @@ static flb_sds_t concat_path(char *p1, char *p2)
+{
+    FILE *fp;
+    int ret;
+
+    fp = fopen(seq_index_file, "r");
+    if (fp == NULL) {
+        return -1;
+    }
+
+    ret = fscanf(fp, "%"PRIu64, seq_index);
+    if (ret != 1) {
+        return -1;
+    }
+
+    fclose(fp);
+    return 0;
+}
+
+/* Writes index value to metadata file */
+static int write_seq_index(char *seq_index_file, uint64_t seq_index)
+{
+    FILE *fp;
+    int ret;
+
+    fp = fopen(seq_index_file, "w+");
+    if (fp == NULL) {
+        return -1;
+    }
+
+    ret = fprintf(fp, "%"PRIu64, seq_index);
+    if (ret < 0) {
+        return -1;
+    }
+
+    fclose(fp);
+    return 0;
+}
+
+static int init_seq_index(void *context) {
+    int ret;
+    int len;
+    const char *tmp;
+    char *tmp_buf;
+    struct flb_s3 *ctx = context;
+
+    ctx->key_fmt_has_seq_index = FLB_TRUE;
+
+    /* Initialize local storage to store metadata files */
+    ret = s3_store_init(ctx);
+    if (ret == -1) {
+        flb_plg_error(ctx->ins, "Failed to initialize S3 storage: %s",
+                      ctx->store_dir);
+        return -1;
+    }
+
+    ctx->stream_metadata = flb_fstore_stream_create(ctx->fs, "metadata");
+    if (!ctx->stream_metadata) {
+        flb_plg_error(ctx->ins, "could not initialize metadata stream");
+        flb_fstore_destroy(ctx->fs);
+        ctx->fs = NULL;
+        return -1;
+    }
+
+    /* Construct directories and file path names */
+    tmp = ctx->stream_metadata->path;
+    ctx->metadata_dir = flb_sds_create(tmp);
+    if (ctx->metadata_dir == NULL) {
+        flb_plg_error(ctx->ins, "Failed to create metadata path");
+        flb_errno();
+        return -1;
+    }
+    tmp = "/index_metadata";
+    ctx->metadata_dir = flb_sds_cat(ctx->metadata_dir, tmp, strlen(tmp));
+    if (ctx->metadata_dir == NULL) {
+        flb_plg_error(ctx->ins, "Failed to create metadata path");
+        flb_errno();
+        return -1;
+    }
+
+    ctx->seq_index_file = flb_sds_create(ctx->metadata_dir);
+    if (ctx->seq_index_file == NULL) {
+        flb_plg_error(ctx->ins, "Failed to create sequential index file path");
+        flb_errno();
+        return -1;
+    }
+    tmp = "/seq_index_";
+    ctx->seq_index_file = flb_sds_cat(ctx->seq_index_file, tmp, strlen(tmp));
+    if (ctx->seq_index_file == NULL) {
+        flb_plg_error(ctx->ins, "Failed to create sequential index file path");
+        flb_errno();
+        return -1;
+    }
+    len = snprintf(NULL, 0, "%d", ctx->ins->id);
+    tmp_buf = flb_malloc(len + 1);
+    if (tmp_buf == NULL) {
+        flb_plg_error(ctx->ins, "Could not allocate enough memory for file name");
+    }
+    snprintf(tmp_buf, len + 1, "%d", ctx->ins->id);
+    tmp_buf[len] = '\0';
+    ctx->seq_index_file = flb_sds_cat(ctx->seq_index_file, tmp_buf, strlen(tmp_buf));
+    flb_free(tmp_buf);
+    if (ctx->seq_index_file == NULL) {
+        flb_plg_error(ctx->ins, "Failed to create sequential index file path");
+        flb_errno();
+        return -1;
```

Comment:
you don't need to figure out how long the plugin id will be and malloc a temporary buffer, that's inefficient

you can rely on the fact that plugin IDs are capped at something like 256. 

Even if it gets increased, a reasonable max value is may be ~1000, so you can just have char array in the function which you use for this purpose. You never need more than 5 chars.